tests/apollo/test_skvbc_reconfiguration.py

```diff
@@ -0,0 +1,229 @@
+         """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+        client = bft_network.random_client()
+
+        await client.write(skvbc.write_req([], [], block_id=0, wedge_command=True))
+
+        with trio.fail_after(seconds=60):
+            done = False
+            while done is False:
+                msg = skvbc.get_have_you_stopped_req()
+                rep = await client.read(msg, m_of_n_quorum=bft_client.MofNQuorum.All(client.config, [r for r in range(
+                    bft_network.config.n)]))
+                rsi_rep = client.get_rsi_replies()
+                done = True
+                for r in rsi_rep.values():
+                    if skvbc.parse_rsi_reply(rep, r) == 0:
+                        done = False
+                        break
+
+        await self.validate_stop_on_super_stable_checkpoint(bft_network, skvbc)
+
+    @with_trio
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 7)
+    async def test_semi_manual_upgrade(self, bft_network):
+        """
+             Sends a wedge command and check that the system stops from processing new requests.
+             Note that in this test we assume no failures and synchronized network.
+             The test does the following:
+             1. A client sends a wedge command
+             2. The client then sends a "Have you stopped" read only command such that each replica answers "I have stopped"
+             3. Apollo stops all replicas
+             4. Apollo starts all replicas
+             5. A client verifies that new write requests are being processed
+         """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+        client = bft_network.random_client()
+
+        key, val = await skvbc.write_known_kv()
+        await client.write(skvbc.write_req([], [], block_id=0, wedge_command=True))
+
+        with trio.fail_after(seconds=60):
+            done = False
+            while done is False:
+                await trio.sleep(seconds=1)
+                msg = skvbc.get_have_you_stopped_req()
+                rep = await client.read(msg, m_of_n_quorum=bft_client.MofNQuorum.All(client.config, [r for r in range(
+                    bft_network.config.n)]))
+                rsi_rep = client.get_rsi_replies()
+                done = True
+                for r in rsi_rep.values():
+                    if skvbc.parse_rsi_reply(rep, r) == 0:
+                        done = False
+                        break
+
+        bft_network.stop_all_replicas()
+
+        # Here the system operator runs a manual upgrade
+
+        bft_network.start_all_replicas()
+
+        await skvbc.assert_kv_write_executed(key, val)
+        await skvbc.write_known_kv()
+
+    async def validate_stop_on_super_stable_checkpoint(self, bft_network, skvbc):
+        for replica_id in range(bft_network.config.n):
+            with trio.fail_after(seconds=90):
```

Comment:
Rather than giving each replica 90 seconds, I think you should give them all 90 seconds or maybe 120 total. You can do this by moving the `trio.fail_after` above the for loop.

Reply:
done